serde_v8/src/ser.rs

```diff
@@ -0,0 +1,649 @@
+    Ok(())
+  }
+
+  fn end(self) -> JsResult<'a> {
+    Ok(self.v8_value.unwrap())
+  }
+}
+
+// Dispatches between magic and regular struct serializers
+pub enum StructSerializers<'a, 'b> {
+  Magic(MagicSerializer<'a, 'b>),
+  Regular(ObjectSerializer<'a, 'b>),
+}
+
+impl<'a, 'b> ser::SerializeStruct for StructSerializers<'a, 'b> {
+  type Ok = JsValue<'a>;
+  type Error = Error;
+
+  fn serialize_field<T: ?Sized + Serialize>(
+    &mut self,
+    key: &'static str,
+    value: &T,
+  ) -> Result<()> {
+    match self {
+      StructSerializers::Magic(s) => s.serialize_field(key, value),
+      StructSerializers::Regular(s) => s.serialize_field(key, value),
+    }
+  }
+
+  fn end(self) -> JsResult<'a> {
+    match self {
+      StructSerializers::Magic(s) => s.end(),
+      StructSerializers::Regular(s) => s.end(),
+    }
+  }
+}
+
+// Serializes to JS Objects, NOT JS Maps ...
+pub struct MapSerializer<'a, 'b> {
+  scope: ScopePtr<'a, 'b>,
+  obj: v8::Local<'a, v8::Object>,
+  next_key: Option<JsValue<'a>>,
+}
+
+impl<'a, 'b> MapSerializer<'a, 'b> {
+  pub fn new(scope: ScopePtr<'a, 'b>) -> Self {
+    let obj = v8::Object::new(&mut *scope.borrow_mut());
+    Self {
+      scope,
+      obj,
+      next_key: None,
+    }
+  }
+}
+
+impl<'a, 'b> ser::SerializeMap for MapSerializer<'a, 'b> {
+  type Ok = JsValue<'a>;
+  type Error = Error;
+
+  fn serialize_key<T: ?Sized + Serialize>(&mut self, key: &T) -> Result<()> {
+    debug_assert!(self.next_key.is_none());
+    self.next_key = Some(key.serialize(Serializer::new(self.scope.clone()))?);
+    Ok(())
+  }
+
+  fn serialize_value<T: ?Sized + Serialize>(
+    &mut self,
+    value: &T,
+  ) -> Result<()> {
+    let v8_value = value.serialize(Serializer::new(self.scope.clone()))?;
+    let scope = &mut *self.scope.borrow_mut();
+    self.obj.set(scope, self.next_key.take().unwrap(), v8_value);
+    Ok(())
+  }
+
+  fn end(self) -> JsResult<'a> {
+    debug_assert!(self.next_key.is_none());
+    Ok(self.obj.into())
+  }
+}
+
+#[derive(Clone)]
+pub struct Serializer<'a, 'b> {
+  scope: ScopePtr<'a, 'b>,
+}
+
+impl<'a, 'b> Serializer<'a, 'b> {
+  pub fn new(scope: ScopePtr<'a, 'b>) -> Self {
+    Serializer { scope }
+  }
+}
+
+macro_rules! forward_to {
+    ($($name:ident($ty:ty, $to:ident, $lt:lifetime);)*) => {
+        $(fn $name(self, v: $ty) -> JsResult<$lt> {
+            self.$to(v as _)
+        })*
+    };
+}
+
+impl<'a, 'b> ser::Serializer for Serializer<'a, 'b> {
+  type Ok = v8::Local<'a, v8::Value>;
+  type Error = Error;
+
+  type SerializeSeq = ArraySerializer<'a, 'b>;
+  type SerializeTuple = ArraySerializer<'a, 'b>;
+  type SerializeTupleStruct = ArraySerializer<'a, 'b>;
+  type SerializeTupleVariant =
+    VariantSerializer<'a, 'b, ArraySerializer<'a, 'b>>;
+  type SerializeMap = MapSerializer<'a, 'b>;
+  type SerializeStruct = StructSerializers<'a, 'b>;
+  type SerializeStructVariant =
+    VariantSerializer<'a, 'b, StructSerializers<'a, 'b>>;
+
+  forward_to! {
```

Comment:
Ditto, could you please remove this macro in favor of repeating method defintions?